Modules/_abc.c

```diff
@@ -0,0 +1,1049 @@
+    if (!PyArg_UnpackTuple(args, "_abc_init", 1, 1, &self)) {
+        return NULL;
+    }
+
+    if (compute_abstract_methods(self) < 0) {
+        return NULL;
+    }
+
+    /* Set up inheritance registry. */
+    data = abc_data_new(&_abc_data_type, NULL, NULL);
+    if (!data) {
+        return NULL;
+    }
+    if (_PyObject_SetAttrId(self, &PyId__abc_impl, data) < 0) {
+        return NULL;
+    }
+    Py_DECREF(data);
+    Py_RETURN_NONE;
+}
+
+PyDoc_STRVAR(_abc_register_doc,
+"Internal ABC helper for subclasss registration. Should be never used outside abc module");
+
+static PyObject *
+_abc_register(PyObject *m, PyObject *args)
+{
+    PyObject *self, *one, *subclass = NULL;
+    int result;
+    if (!PyArg_UnpackTuple(args, "_abc_register", 2, 2, &self, &subclass)) {
+        return NULL;
+    }
+    if (!PyType_Check(subclass)) {
+        PyErr_SetString(PyExc_TypeError, "Can only register classes");
+        return NULL;
+    }
+    result = PyObject_IsSubclass(subclass, self);
+    if (result > 0) {
+        Py_INCREF(subclass);
+        return subclass;  /* Already a subclass. */
+    }
+    if (result < 0) {
+        return NULL;
+    }
+    /* Subtle: test for cycles *after* testing for "already a subclass";
+       this means we allow X.register(X) and interpret it as a no-op. */
+    result = PyObject_IsSubclass(self, subclass);
+    if (result > 0) {
+        /* This would create a cycle, which is bad for the algorithm below. */
+        PyErr_SetString(PyExc_RuntimeError, "Refusing to create an inheritance cycle");
+        return NULL;
+    }
+    if (result < 0) {
+        return NULL;
+    }
+    if (_add_to_registry(self, subclass) < 0) {
+        return NULL;
+    }
+    Py_INCREF(subclass);
+    /* Invalidate negative cache */
+    Py_DECREF(abc_invalidation_counter);
+    one = PyLong_FromLong(1);
+    abc_invalidation_counter = PyNumber_Add(abc_invalidation_counter, one);
+    Py_DECREF(one);
+    return subclass;
+}
+
+PyDoc_STRVAR(_abc_instancecheck_doc,
+"Internal ABC helper for instance checks. Should be never used outside abc module");
+
+static PyObject *
+_abc_instancecheck(PyObject *m, PyObject *args)
+{
+    PyObject *self, *result, *subclass, *subtype, *instance = NULL;
+    int incache;
+    if (!PyArg_UnpackTuple(args, "_abc_instancecheck", 2, 2, &self, &instance)) {
+        return NULL;
+    }
+    subclass = _PyObject_GetAttrId(instance, &PyId___class__);
+    /* Inline the cache checking. */
+    incache = _in_cache(self, subclass);
+    if (incache < 0) {
+        Py_DECREF(subclass);
+        return NULL;
+    }
+    if (incache > 0) {
+        Py_DECREF(subclass);
+        Py_RETURN_TRUE;
+    }
+    subtype = (PyObject *)Py_TYPE(instance);
+    if (subtype == subclass) {
+        incache = _in_negative_cache(self, subclass);
+        if (incache < 0) {
+            Py_DECREF(subclass);
+            return NULL;
+        }
+        if ((PyObject_RichCompareBool(_get_negative_cache_version(self),
```

Comment:
`_get_negative_cache_version(self)` needs to be called separately, and it's result compared to `NULL` explicitly to bail out early. This code isn't correct.

Reply:
`PyObject_RichCompareBool` can return `-1` -- error that isn't handled right now.